client/async.go

```diff
@@ -0,0 +1,266 @@
+	ctx context.Context,
+	method string,
+	req interface{},
+	reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	l := a.logger.WithFields(logrus.Fields{
+		"method": method,
+	})
+
+	events := req.(*pb.SendEventsRequest).Events
+
+	defer func(startTime time.Time) {
+		elapsedTime := float64(time.Since(startTime).Nanoseconds() / 1000000)
+		for _, e := range events {
+			metrics.ClientRequestsResponseTime.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+			).Observe(elapsedTime)
+		}
+		l.WithFields(logrus.Fields{
+			"elapsedTime": elapsedTime,
+			"reply":       reply.(*pb.SendEventsResponse),
+		}).Debug("request processed")
+	}(time.Now())
+
+	if err := invoker(ctx, method, req, reply, cc, opts...); err != nil {
+		l.WithError(err).Error("error processing request")
+		for _, e := range events {
+			metrics.ClientRequestsFailureCounter.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+				err.Error(),
+			).Inc()
+		}
+	}
+	for _, e := range events {
+		metrics.ClientRequestsSuccessCounter.WithLabelValues(
+			hostname,
+			method,
+			e.Topic,
+		).Inc()
+	}
+	return nil
+}
+
+func (a *gRPCClientAsync) send(ctx context.Context, event *pb.Event) error {
+	a.wg.Add(1)
+	a.eventsChannel <- event
+	return nil
+}
+
+func (a *gRPCClientAsync) sendRoutine() {
+	ticker := time.NewTicker(a.flushInterval)
+	defer ticker.Stop()
+
+	req := &pb.SendEventsRequest{}
+	req.Events = make([]*pb.Event, 0, a.flushSize)
+
+	send := func() {
+		cpy := req
+		cpy.Id = uuid.NewV4().String()
+		req = &pb.SendEventsRequest{}
+		req.Events = make([]*pb.Event, 0, a.flushSize)
+		go a.sendEvents(cpy, 0)
+	}
+
+	for {
+		select {
+		case e := <-a.eventsChannel:
+			if len(req.Events) == 0 {
+				a.wg.Add(1)
+			}
+			a.wg.Done()
+			req.Events = append(req.Events, e)
+			if len(req.Events) == a.flushSize {
+				send()
+			}
+		case <-ticker.C:
+			if len(req.Events) > 0 {
+				send()
+			}
+		}
+	}
+}
+
+func (a *gRPCClientAsync) sendEvents(req *pb.SendEventsRequest, retryCount int) {
+	l := a.logger.WithFields(logrus.Fields{
+		"operation":  "sendEvents",
+		"requestId":  req.Id,
+		"retryCount": retryCount,
+		"size":       len(req.Events),
+	})
+	if retryCount > a.maxRetries {
+		l.Info("dropped events due to max retries")
+		a.wg.Done()
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), a.timeout)
+	defer cancel()
+	// in case server's producer fail to send any event, failure indexes are sent
+	// in response to be retried
+	res, err := a.client.SendEvents(ctx, req)
+	if err != nil {
+		l.WithError(err).Error("failed to send events")
+		time.Sleep(time.Duration(math.Pow(2, float64(retryCount))) * a.retryInterval)
+		a.sendEvents(req, retryCount+1)
+		return
+	}
+	if res.FailureIndexes != nil && len(res.FailureIndexes) != 0 {
+		l.WithFields(logrus.Fields{
+			"failureIndexes": res.FailureIndexes,
+		}).Error("failed to send events")
```

Comment:
Besides logging, are we keeping track of retried events? I'd say we send a tag with a bool `retry` when we report that an event was sent.

Reply:
nice! will add to metrics and proto